ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/CommandTopicBackup.java

```diff
@@ -0,0 +1,210 @@
+ * complete backup of the command_topic is created.
+ */
+public class CommandTopicBackup {
+  private static final Logger LOG = LoggerFactory.getLogger(CommandTopicBackup.class);
+  private static final Ticker CURRENT_MILLIS_TICKER = new Ticker() {
+    @Override
+    public long read() {
+      return System.currentTimeMillis();
+    }
+  };
+  private static final String PREFIX = "backup_";
+
+  private final File backupLocation;
+  private final String topicName;
+  private final Ticker ticker;
+
+  private BackupReplayFile replayFile;
+  private List<Pair<CommandId, Command>> latestReplay;
+  private int latestReplayIdx;
+
+  public CommandTopicBackup(final String location, final String topicName) {
+    this(location, topicName, CURRENT_MILLIS_TICKER);
+  }
+
+  public CommandTopicBackup(final String location, final String topicName, final Ticker ticker) {
+    final File dir = new File(Objects.requireNonNull(location, "location"));
+    if (!dir.exists() || !dir.isDirectory()) {
+      throw new KsqlException(String.format(
+          "Backup location '%s' does not exist or it is not a directory.", location));
+    }
+
+    this.backupLocation = dir;
+    this.topicName = Objects.requireNonNull(topicName, "topicName");
+    this.ticker = Objects.requireNonNull(ticker, "ticker");
+  }
+
+  public void initialize() {
+    replayFile = openOrCreateReplayFile();
+
+    try {
+      latestReplay = replayFile.readRecords();
+    } catch (final IOException e) {
+      LOG.warn("Failed to read the latest backup from {}. Continue with a new file. Error = {}",
+          replayFile.getPath(), e.getMessage());
+
+      replayFile = newReplayFile();
+      latestReplay = Collections.emptyList();
+    }
+
+    latestReplayIdx = 0;
+    LOG.info("Command topic will be backup on file: {}", replayFile.getPath());
+  }
+
+  public void close() {
+    try {
+      replayFile.close();
+    } catch (final IOException e) {
+      LOG.warn("Failed closing the backup file {}. Error = {}",
+          replayFile.getPath(), e.getMessage());
+    }
+  }
+
+  @VisibleForTesting
+  BackupReplayFile getReplayFile() {
+    return replayFile;
+  }
+
+  public void writeRecord(final ConsumerRecord<CommandId, Command> record) {
+    if (latestReplayIdx < latestReplay.size()) {
+      final Pair<CommandId, Command> latestReplayRecord = latestReplay.get(latestReplayIdx);
+      if (record.key().equals(latestReplayRecord.left)
+          && record.value().equals(latestReplayRecord.right)) {
+        // Ignore backup because record was already replayed
+        latestReplayIdx++;
+        return;
+      } else {
+        LOG.info("Previous command topic backup does not match the new command topic data. "
+            + "A new backup file will be created.");
+        createNewBackupFile();
+        LOG.info("New backup file created: {}", replayFile.getPath());
+      }
+    } else if (latestReplayIdx > 0) {
+      // clear latest replay from memory
+      latestReplay.clear();
```

Comment:
`Collections.emptyList` will throw an exception if `clear` gets called on it. While I don't think it's possible for this situation to be hit because `latestReplayIdx` should never be greater than `latestReplay.size()` we might want to change this to `latestReaply.size() > 0` instead

Reply:
Done.